backends/hip-gen/ceed-hip-gen-qfunction.c

```diff
@@ -124,15 +61,19 @@ int CeedQFunctionCreate_Hip_gen(CeedQFunction qf) {
   ierr = CeedCalloc(1, &data); CeedChkBackend(ierr);
   ierr = CeedQFunctionSetData(qf, data); CeedChkBackend(ierr);
 
-  char *source;
-  ierr = CeedQFunctionGetSourcePath(qf, &source); CeedChkBackend(ierr);
-  const char *funname = strrchr(source, ':') + 1;
-  data->qFunctionName = (char *)funname;
-  const int filenamelen = funname - source;
-  char filename[filenamelen];
-  memcpy(filename, source, filenamelen - 1);
-  filename[filenamelen - 1] = '\0';
-  ierr = loadHipFunction(qf, filename); CeedChkBackend(ierr);
+  char *source_path;
+  ierr = CeedQFunctionGetSourcePath(qf, &source_path); CeedChkBackend(ierr);
+  if (source_path[0] != '\0') {
```

Comment:
You forgot the documentation, as a result the code becomes obscure -> function.

Reply:
Which part did I forget the documentation of? `CeedQFunctionGetSourcePath()` is documented. `CeedQFunctionGetKernelName()` is documented. `CeedLoadSourceToBuffer()` is documented.

Reply:
```c++
if (source_path[0] != '\0') {
```
usually you have this comment above:
```c++
// Empty source path indicates user must supply Q-Function
```

Reply:
That comment isn't applicable in the HIP/CUDA gen backends though. We must error if no QFunction source is provided. I suspect that in the HIP/CUDA ref backends we don't actually need the `else` clause though.

Reply:
That's even more reasons to centralize this logic into a function.

Reply:
I don't understand. We do two different things in four different places so we should make a function that does these two different things in four different places?

Instead of just saying 'make it into a function' can you please provide more details? Right now the logic is just 'if there is a QFunction source path, then call these two functions' and for two code gen backends there is the extra clause 'if there is not, then error'.

Reply:
> I don't understand. We do two different things in four different places so we should make a function that does these two different things in four different places?

Yes, the idea is to have more self-documented code, currently when I read your code I see __how__ you do things, but I don't know __what__ you do, I have to deduce the "what" from the "how". Functions allow you to separate level of abstractions so that when I read code I stay at the same level of abstraction, understand "what"it does without me having to know implementation details, and you don't mix low level code (ex: `source_path[0] != '\0'`) with high level code. The function names should suggest their high level intent (the "what"). You can either make one function that does it all if the details don't matter, if you think the distinction between the two different processing matters, then you can write two functions that convey the difference in the "what".

> Instead of just saying 'make it into a function' can you please provide more details? Right now the logic is just 'if there is a QFunction source path, then call these two functions' and for two code gen backends there is the extra clause 'if there is not, then error'.

I think the important part here is to find "what" it does, your "logic" only describes the "how".

Reply:
This is C, so to some degree we are are stuck using some low level approaches. Is there a different way to check if the source file path string is empty that you would prefer?

In this chunk

```c
// Read source
char *source_path;
ierr = CeedQFunctionGetSourcePath(qf, &source_path); CeedChkBackend(ierr);
// Empty source path indicates user must supply Q-Function
if (source_path[0] != '\0') {
  ierr = CeedQFunctionGetKernelName(qf, &data->qFunctionName); CeedChkBackend(ierr);
  ierr = CeedLoadSourceToBuffer(ceed, source_path, &data->qFunctionSource); CeedChkBackend(ierr);
}
```

we are loading the kernel name and source file contents into the correct buffer for the implementation dependent object. We currently have 4 different objects being populated with this information, CUDA ref QFunction backend data, CUDA gen QFunction backend data, HIP ref QFunction data, and HIP gen QFunction backend data. I don't see an obvious clean way to make one function that takes in these four different function definitions.

I could wrap `CeedLoadSourceToBuffer()` in `CeedQFunctionLoadSourceToBuffer()` if you think it would be clearer. We could then check if that string is empty and error if the source buffer is empty for the code generation backends. That would get us down to 2 function calls in 4 places and 2 if statements in 2 places.

Side note: This is code you wrote that has been modified since that time and then copied into the HIP backend. When making changes to the GPU backends, I try to be somewhat conservative in my changes since I am not the primary maintainer of the GPU backends.

Reply:
Ok, how about this pattern:
```c
ierr = CeedQFunctionGetKernelName(qf, &data->qFunctionName); CeedChkBackend(ierr);
ierr = CeedQFunctionLoadSourceToBuffer(qf, &data->qFunctionSource); CeedChkBackend(ierr);
```

Reply:
I agree that the code I wrote for the cuda-gen backend years ago is difficult to read, difficult to maintain, and difficult to extend, and overall not a nice looking code. I have to say that it was a pain for me to write this in C, I had to mimic C++ generic template programming using C-strings, that was not fun. Just a reminder that this code was written to get performance on GPU which wasn't the case before for libCEED, the ambition was not to rewrite all the little things that I find badly written. I think though that once code has been reviewed and accepted we should not consider that the author is the only one responsible for maintaining it, the purpose of reviewing is also to make sure that the code is good enough so that pretty much __anyone__ in the project can read, maintain, and extend it. I don't think it's completely fair to complain on the quality of the code I wrote which the team reviewed (reviewers have their share of responsibility), and it should absolutely not be an excuse to write bad code now.

This PR is 6500 lines diff, 106 files modified, and most of this diff is duplicated code, as a result it's relatively a pain to review, and it feels like a programming version of the 7 errors game.

This being said, I think that having this kind of code is an anti-pattern to me:
```c
if (source_path[0] != '\0') {
   ierr = CeedLoadSourceToBuffer(ceed, source_path, &data->qFunctionSource); CeedChkBackend(ierr);
} else {
   return CeedError(ceed, CEED_ERROR_UNSUPPORTED,
                  "/gpu/hip/gen backend requires QFunction source code file");
}
```
This means that a user/developer __has to__ be aware of this special semantic of the `source_path` in order to use `CeedQFunctionLoadSourceToBuffer` properly, and there is no reason for that. This ` if` could be inside `CeedQFunctionLoadSourceToBuffer`, and the code that access the ` source_path` could also be inside `CeedQFunctionLoadSourceToBuffer`. If you want to customize the error message you can add a parameter. Then for the case that is different, you can either have a different behavior when no message is given, and document the semantic of it, or you create a function with a different name built on top of the previous function that has the expected different behavior.
By doing this you indeed obtain something looking like your previous answer, the benefit is that I directly understand __what__ is happening, and I don't have to even know __how__ it's done. Another benefit is that all the logic about the __how__ is in one place, reducing significantly the risk of errors when refactoring, and also improving readability, maintainability, and safety. In many aspects this PR is refactoring code, and for instance I cannot know as is if you didn't omit to modify this pattern somewhere else in the code.

Reply:
I think you are now commented about a pattern I have already removed. Did you miss the update that I put in? Do you like the update I put in? Are you requesting changes to the update I put in?

I agree with your comments about the sprawling nature of this PR. This is exactly why I think we should be making a larger number of smaller PRs. I would love to see a full refactor of the GPU backends, but that should be done in a series of smaller PRs.

To be honest, I don't know enough about the design of the GPU backends at this time to feel comfortable considering myself a primary maintainer of those backends. I'm in the awkward position of pushing forward library features into code that others have designed. It really isn't reasonable to omit GPU implementations of these features so I'm tinkering with backend code I don't really feel like I have ownership over. It is a *very* frustrating situation that I have been trying to remedy by learning about and documenting the GPU backend code as I have time, but less hostile code review would be helpful.

Reply:
I'm sorry that me telling you how I would improve this work makes you so emotional, this is not my intention, and I apologize for that. I already have spent a lot of time on this PR, and I really have much more important priorities right now, I don't have the time to micro-manage this work. Please ask me to review again your work when you feel confident that it's in a good state.

If you want me to explain anything to you about code that I wrote I'm always available for that. You can also open issues for documentation requests.

Reply:
It's a lot of lines of code touched, but it is half the size.
```
106 files changed, 1812 insertions(+), 3815 deletions(-)
```
I'd hope we can lean on tests so that the reviewing need is mostly for whether this is boxing us into a position where future features (including performance optimization) will require excessive churn. If this is an improvement, but the concept can be pushed further, I think it's reasonable to approve and leave the next increment for another PR, which would hopefully be quite a bit smaller. We really appreciate your time reviewing -- I'll note that Jeremy also has other deliverables (for PSAAP, which funds him now). I think we should have a long-term governance and sustainability discussion sometime in the coming months.

On the specific issue of strings, I'm not a fan of the empty string being special. A pointer that can be NULL is like `Option<String>`, okay, and sometimes we need a `String`. But empty string (path) is in my book like any path that doesn't exist -- we can try to open it and give a reasonable error if it fails.

Reply:
I was actually explicitly `NULL`ing the pointer to the string buffer in `CeedQFunctionLoadSourceToBuffer()` but didn't use that in my error checking. I swapped that in instead of using `strlen` since it is a bit clearer (though the effect is identical for a `NULL` pointer)
```C
// Read QFunction source
ierr = CeedQFunctionGetKernelName(qf, &data->qFunctionName); CeedChkBackend(ierr);
ierr = CeedQFunctionLoadSourceToBuffer(qf, &data->qFunctionSource); CeedChkBackend(ierr);
if (!data->qFunctionSource)
  return CeedError(ceed, CEED_ERROR_UNSUPPORTED,
                   "/gpu/cuda/gen backend requires QFunction source code file");
```
Those 2/3 statements seem really clear to me now.